Add support for adding RIDs to runtime.json during build

ericstj opened the pull request:
Fixes #48507

cc @omajid 

This replaces https://github.com/dotnet/arcade/pull/7141 and https://github.com/dotnet/runtime/pull/50157

The most interesting thing to look at are the tests for GenerateRuntimeGraph, those show how RID additions are handled.  It took a bit of refinement until I could come up with something that I think was most consistent with what would happen if we actually made source changes to include the RIDs.

ericstj committed:
Add support for adding RIDs to RidGraph during build

safern commented on src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj:
```diff
@@ -38,5 +48,14 @@
     <PackageReference Include="NuGet.ProjectModel" Version="$(RefOnlyNugetProjectModelVersion)" />
   </ItemGroup>
 
+  <Target Name="GenerateRuntimeJson" Condition="'$(AdditionalRuntimeIdentifiers)' != ''" BeforeTargets="Pack">
```
Should this have inputs and outputs?

ericstj replied:
It runs on pack and is meant to only run when properties are set (not easily tracked by inputs/outputs) so I intentionally made it always run.

safern commented on src/libraries/Microsoft.NETCore.Platforms/src/RID.cs:
```diff
@@ -40,20 +47,153 @@ public override string ToString()
+                        else if (current == ArchitectureDelimiter)
+                        {
+                            // ensure there's no version later in the string
+                            if (-1 != runtimeIdentifier.IndexOf(VersionDelimiter, i))
```
this feels weird to rid when the `-1` is at the beginning.

ericstj replied:
I once worked on a team where they always put constants on the left since it would catch accidental assignment. I’m fine changing it.

safern commented on src/libraries/Microsoft.NETCore.Platforms/src/RID.cs:
```diff
@@ -40,20 +47,153 @@ public override string ToString()
+        {
+            Base = 0,
+            Version,
+            Architcture,
```
```suggestion
            Architecture,
```

ericstj replied:
Dangerous suggestion,  need to make the change everywhere :)

safern replied:
Yeah I realized afterwards, but for typos like this the suggestion displays the diff better 🙈

ericstj replied:
No worries, but you don't get the commit credit 😆

safern commented on src/libraries/Microsoft.NETCore.Platforms/src/RID.cs:
```diff
@@ -40,20 +47,153 @@ public override string ToString()
+            {
+                if (partLength == 0)
+                {
+                    throw new ArgumentException($"unexpected delimiter at position {partStart} in {runtimeIdentifier}");
```
Should we include what we expected and the delimeter we found?

ericstj replied:
The entire rid is already included. I think it’s more useful to look at the entire string rather than a single character. It’s wrong in context not on its own.

safern replied:
Ah you are right I missed that the runtime identified it missed to parse is included

ericstj replied:
I'll quote it to make it obvious

safern commented on src/libraries/Microsoft.NETCore.Platforms/src/RID.cs:
```diff
@@ -40,20 +47,153 @@ public override string ToString()
         {
-            return Architecture != null;
+            return object.ReferenceEquals(obj, this) ||
+                (!(obj is null) &&
```
can we use `is not null`? 
```suggestion
                (obj is not null &&
```

ericstj replied:
An opportunity for https://github.com/dotnet/runtime/blob/f6b344b64a367ad71df11782122716b2e616d88f/src/libraries/System.Private.CoreLib/src/System/Version.cs#L159 where I copied this from.

safern commented on src/libraries/Microsoft.NETCore.Platforms/tests/GenerateRuntimeGraphTests.cs:
```diff
@@ -21,38 +23,219 @@ public GenerateRuntimeGraphTests(ITestOutputHelper output)
         {
-            TaskItem item = new TaskItem(projectItem.EvaluatedInclude);
-            foreach(var metadatum in projectItem.Metadata)
+            // will generate and compare to existing file.
```
Should we assert that the generated file has the additional rids?

safern replied:
Ah, this test adds rids that already exist?

ericstj replied:
Right. This test passes if the addition of RIDs make no change to the checked in runtime.json

safern commented on src/libraries/Microsoft.NETCore.Platforms/tests/GenerateRuntimeGraphTests.cs:
```diff
@@ -21,38 +23,219 @@ public GenerateRuntimeGraphTests(ITestOutputHelper output)
+        /// <param name="runtimeFilePrefix">a unique prefix to use for the generated </param>
+        private void AssertRuntimeGraphAdditions(string[] additionalRIDs, RuntimeDescription[] expectedAdditions, string additionalRIDParent = null, [CallerMemberName] string runtimeFilePrefix = null)
+        {
+            string runtimeFile = runtimeFilePrefix + ".runtime.json";
```
Should these files be created on a temp directory and cleaned up after the test class is disposed? 

We have: https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/TestUtilities/System/IO/FileCleanupTestBase.cs to achieve that.

ericstj replied:
Test working directory is a temp directory. I don’t think cleaning them up is desirable, they are useful for debugging.

safern replied:
Ok, as long as the working dir is a temp directory, I'm fine with it not being cleaned up.

ericstj replied:
I'm calling it temp in the context of it's the test output directory.  It will be cleaned up by normal dev activities.  I intentionally don't want to hide these files in some random location as they are useful for diffing what the result of the test is.

ericstj committed:
Address code review feedback

tmds commented on src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj:
```diff
@@ -14,6 +14,13 @@
+    <!-- When building from source, ensure the RID we're building for is part of the RID graph -->
+    <AdditionalRuntimeIdentifiers>$(AdditionalRuntimeIdentifiers);$(OutputRID)</AdditionalRuntimeIdentifiers>
+    <!-- Also ensure the RID we're bulding on is part of the RID graph, since this may be more specific -->
+    <AdditionalRuntimeIdentifiers>$(AdditionalRuntimeIdentifiers);$([System.Runtime.InteropServices.RuntimeInformation]::RuntimeIdentifier)</AdditionalRuntimeIdentifiers>
```
I'm not sure we need to add the second.

`OutputRID` should be `$([System.Runtime.InteropServices.RuntimeInformation]::RuntimeIdentifier)` that ignores `DOTNET_RUNTIME_ID` envvar.

ericstj replied:
>`$([System.Runtime.InteropServices.RuntimeInformation]::RuntimeIdentifier)` that ignores `DOTNET_RUNTIME_ID` envvar

`RuntimeInformation.RuntimeIdentifier` does honor the `DOTNET_RUNTIME_ID` envvar.  What did you mean by ignores?

I added both based on your suggestion in the issue.  I think having both might better support a case where someone is building portable Linux from source (OutputRid=linux-x64) and expecting it to support a new distro (myLinux.1.0-x64)

tmds replied:
>  What did you mean by ignores?

I assume (but maybe this is wrong) that it may be necessary to set `DOTNET_RUNTIME_ID` when building the source-built tarball on a distribution that is not in the graph of the SDK used during the build.
In this case, we'd still want the `OutputRid` to reflect `/etc/os-release`. We'd be setting `DOTNET_RUNTIME_ID` just to make the build work.

ericstj replied:
I see what you mean.  This is a bit of chicken and egg.

How do we even run [the code](https://github.com/dotnet/runtime/blob/3df7a6e3dd4ab11494f4e3a2394b386a73bb446f/src/native/corehost/hostmisc/utils.cpp#L203-L220) to [read from](https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/native/corehost/hostmisc/pal.unix.cpp#L533-L796)`/etc/os-release` to calculate `OutputRID` if that code needs `DOTNET_RUNTIME_ID` set in order to run?  The host needs to "know" what RID it's running on so that it can make decisions about which RID-specific assets to select, and its the same mechanism that impacts [this API](https://github.com/dotnet/runtime/issues/26780).

We could just make the user specify `OutputRID` on the command-line, but that's lame and requires the user to get it right.

If we didn't want that, we'd need someway to run this bit of code from the host and use it during the build, instead of `$([System.Runtime.InteropServices.RuntimeInformation]::RuntimeIdentifier)` as is done today.  This needs to happen outside this task/project since it's something that the build needs to consume (when producing RID-specific packages).  

@agocke @vitek-karas @VSadov what do you think about this: how should we best automatically calculate the "ideal" RID for our purposes of source-build on new platforms.

ericstj replied:
Another thought: should you even need to set `DOTNET_RUNTIME_ID`?  Ideally in the case where a `linux-arch` host build is running on a machine that's say `someDistro.version.arch` it should use `someDistro.version.arch` if that's "known", but if it isn't "known" seems like the host can just assume `linux-arch` for resolving assets (since it's running after all).  Is this how the host works?  Should it?

tmds replied:
> should you even need to set DOTNET_RUNTIME_ID?

Maybe this will just work, I'm not sure. Let's assume it does.

Once this is merged, we should add a test in source-build repo that verifies this use-case works.
The test can change `/etc/os-release` to something unknown and see if things still build.

vitek-karas replied:
The host has relatively complicated code to determine the current RID: https://github.com/dotnet/runtime/blob/bc9f00c226c7716c893576d53ea45a616b2b2c36/src/native/corehost/hostmisc/pal.unix.cpp#L710 (there are special versions of this for certain distros, same source file).
This algorithm may "fail", in which case the host falls back to a basic RID - on Linux it will be simply `linux` RID. https://github.com/dotnet/runtime/blob/bc9f00c226c7716c893576d53ea45a616b2b2c36/src/native/corehost/hostmisc/pal.h#L85

If this algorithm works, then there should be no need for `DOTNET_RUNTIME_ID`. If it fails and the fallback is not enough, or if it returns wrong value, then `DOTNET_RUNTIME_ID` is a way to override it and might be necessary. In that case though, we might want to bake the new RID into the hosts produced by the source build somehow, so that running apps using the results of such source build won't require `DOTNET_RUNTIME_ID` to be set. Not sure what the workflow here is...

ericstj replied:
Ah, very cool.  In that case I think not setting `DOTNET_RUNTIME_ID` should just work.  I'll go ahead and remove the second value that's being added to `AdditionalRuntimeIdentifiers`.  I like it, this simplifies the the behavior here.

>Once this is merged, we should add a test in source-build repo that verifies this use-case works.
The test can change /etc/os-release to something unknown and see if things still build.

Can you help with this part @tmds or @omajid?  We probably also want to make sure that the generated runtime.json will include that new RID (to ensure it's flowing through the targets correctly).

tmds replied:
> Can you help with this part @tmds or @omajid?

@crummel and @MichaelSimons can help with this. I can take a look also, but I'm not familiar with the source-build CI setup.

ericstj committed:
Ensure GenerateRuntimeJson runs before Pack

The pack task itself doesn't do anything but depend on a sequence of tasks.
As a result running a target `BeforeTargets="Pack"` actually runs after
the package is created, since dependencies run before BeforeTargets.

Sequence the target before GenerateNuspec instead.

ericstj committed:
Only add target RID to the RID graph

tmds commented on src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj:
```diff
@@ -12,6 +12,9 @@
     <PackageDescription>Provides runtime information required to resolve target framework, platform, and runtime specific implementations of .NETCore packages.</PackageDescription>
+    
+    <!-- When building from source, ensure the RID we're building for is part of the RID graph -->
+    <AdditionalRuntimeIdentifiers Condition="'$(DotNetBuildFromSource)' == 'true'">$(AdditionalRuntimeIdentifiers);$(OutputRID)</AdditionalRuntimeIdentifiers>
```
It's good we're appending the `OutputRID` at the end here. The `AdditionalRuntimeIdentifiers` will be added in order, so we want them to be ordered from least to most specific.

ericstj replied:
That shouldn’t matter since the only way to establish a parent RID is through the parent parameter.  New RIDs are added recursively following the rules in the runtime groups and parent.  When written the graph is sorted.

tmds reviewed (approved):
LGTM, thanks!

ericstj commented:
@ViktorHofer or @safern can I get a review here?

ViktorHofer commented on src/libraries/Microsoft.NETCore.Platforms/src/RID.cs:
```diff
@@ -1,37 +1,44 @@
 namespace Microsoft.NETCore.Platforms.BuildTasks
 {
-    internal class RID
+    public class RID
```
I guess the access modifier changed because of tests depending on the RID class now?

ericstj replied:
Yeah, I didn't bother setting up internals visible to since this is an msbuild task and the only consumer of it is our build and tests.

ViktorHofer commented on src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj:
```diff
@@ -38,5 +44,14 @@
+    <GenerateRuntimeGraph RuntimeGroups="@(RuntimeGroupWithQualifiers)"
+                          AdditionalRuntimeIdentifiers="$(AdditionalRuntimeIdentifiers)"
+                          AdditionalRuntimeIdentifierParent="$(AdditionalRuntimeIdentifierParent)"
+                          RuntimeJson="$(IntermediateOutputPath)runtime.json"
```
nit: consider defining a property for the path to the runtime.json file. You are currently hardcoding it in two places.

ericstj replied:
will do so in a follow up PR

ViktorHofer reviewed (approved):
Looks great to me

safern commented:
LGTM! Thanks for addressing the questions/comments I had.
